sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -327,3 +345,104 @@ object JoinReorderDP extends PredicateHelper with Logging {
 case class Cost(card: BigInt, size: BigInt) {
   def +(other: Cost): Cost = Cost(this.card + other.card, this.size + other.size)
 }
+
+/**
+ * Implements optional filters to reduce the search space for join enumeration.
+ *
+ * 1) Star-join filters: Plan star-joins together since they are assumed
+ *    to have an optimal execution based on their RI relationship.
+ * 2) Cartesian products: Defer their planning later in the graph to avoid
+ *    large intermediate results (expanding joins, in general).
+ * 3) Composite inners: Don't generate "bushy tree" plans to avoid materializing
+ *   intermediate results.
+ *
+ * Filters (2) and (3) are not implemented.
+ */
+case class JoinReorderDPFilters(conf: SQLConf) extends PredicateHelper {
+  /**
+   * Builds join graph information to be used by the filtering strategies.
+   * Currently, it builds the sets of star/non-star joins.
+   * It can be extended with the sets of connected/unconnected joins, which
+   * can be used to filter Cartesian products.
+   */
+  def buildJoinGraphInfo(
+      items: Seq[LogicalPlan],
+      conditions: Set[Expression],
+      planIndex: Seq[(LogicalPlan, Int)]): Option[JoinGraphInfo] = {
+
+    // Compute the tables in a star-schema relationship.
+    val starJoin = StarSchemaDetection(conf).findStarJoins(items, conditions.toSeq)
+    val nonStarJoin = items.filterNot(starJoin.contains(_))
+
+    if (starJoin.nonEmpty && nonStarJoin.nonEmpty) {
+      val (starInt, nonStarInt) = planIndex.collect {
+        case (p, i) if starJoin.contains(p) =>
+          (Some(i), None)
+        case (p, i) if nonStarJoin.contains(p) =>
+          (None, Some(i))
+        case _ =>
+          (None, None)
+      }.unzip
+      Some(JoinGraphInfo(starInt.flatten.toSet, nonStarInt.flatten.toSet))
+    } else {
+      // Nothing interesting to return.
+      None
+    }
+  }
+
+  /**
+   * Applies star-join filter.
+   *
+   * Given the outer/inner and the star/non-star sets,
+   * the following plan combinations are allowed:
+   * 1) (outer U inner) is a subset of star-join
```

Comment:
@cloud-fan The ```outer/inner``` represents the join plan combinations generated by ```JoinReorderDP```. ```JoinReorderDP``` calls them ```oneJoinPlan/otherJoinPlan```. I will rename them to align to join DP.